### PR TITLE
DAC-238: Abort a certification run

### PR DESCRIPTION
**Pull request by bogdan-manole**
Adding support for aborting a Run

### CODE REVIEW

File: src/Plutus/Certification/Local.hs [hunk 1]
```diff
@@ -67,6 +67,7 @@ localServerCaps backend = do
 
     getRuns _ (RunID jobId) =
       Map.findWithDefault [] jobId <$> (lift $ readIORef jobs) >>= yieldMany
+    abortRuns = error "not implemented"
```
Comment (shlevy) [inline src/Plutus/Certification/Local.hs:70]: Instead of the line marked "Purposefully leak" above, we can store the return from `async` in the state and then kill the async run from here. Processes should be killed, tempdirs cleaned up, etc.

File: src/Plutus/Certification/Server.hs [hunk 2]
```diff
@@ -29,12 +30,15 @@ data ServerCaps m r = ServerCaps
     submitJob :: !(EventBackendModifiers r r -> FlakeRefV1 -> m RunIDV1)
   , -- | Get the status of all runs associated with a job
     getRuns :: !(EventBackendModifiers r r -> RunIDV1 -> ConduitT () RunStatusV1 m ())
+  , -- | Delete all runs associated with a job
+    abortRuns :: !(EventBackendModifiers r r -> RunIDV1 -> ConduitT () () m ())
```
Comment (shlevy) [inline src/Plutus/Certification/Server.hs:34]: Why return a conduit here? Why not just `m ()`?

File: src/Plutus/Certification/Server.hs [hunk 3]
```diff
@@ -68,6 +74,8 @@ server ServerCaps {..} eb = NamedAPI
      runConduit
         $ getRuns (setAncestor $ reference ev) rid
        .| evalStateC Queued consumeRuns
+  , abortRun = \rid -> withEvent eb AbortRun \ev -> do
```
Comment (shlevy) [inline src/Plutus/Certification/Server.hs:77]: Based on the type you gave to the `AbortRun` constructor it looks like you intended to add the runId to the event?

**Review by shlevy — COMMENTED** [review]
Looking good overall! A few comments.

File: src/Plutus/Certification/Local.hs (hunk 1 above)
Comment (bogdan-manole) [reply src/Plutus/Certification/Local.hs]: done.

NOTE: I added another `IORef` for cancellations. The other two options would be to use the same `IORef`, but  IMO the code looks much more straightforward.  Do you think we should clean the Map entry after the job is finished?

File: src/Plutus/Certification/Cicero.hs [hunk 4]
```diff
@@ -105,6 +107,22 @@ ciceroServerCaps backend CiceroCaps {..} = ServerCaps {..}
...
+    abortRuns mods rid = go 0
+      where
+        eb = modifyEventBackend mods backend
+        rid' = Cicero.Fact.FactID $ rid.uuid
+        limit = 10
+        go offset = do
+          runs <- lift . runClientOrDie clientCaps eb $ ciceroClient.run.getAll True [rid'] (Just offset) (Just limit)
+          let count = List.genericLength runs
+          yieldMany runs
+            .| filterC (isNothing . (.finishedAt))
```
Comment (shlevy) [inline src/Plutus/Certification/Cicero.hs:119]: Note that in principle anyone could set up arbitrary cicero jobs downstream of our facts. But I don't think this will be an issue and I'd like to get off of cicero anyway.

File: src/Plutus/Certification/Local.hs (hunk 1 above)
Comment (shlevy) [reply src/Plutus/Certification/Local.hs]: Can't clean the status map entry since you need to check the status after the job is done, but yeah should be safe to remove the `Async` afterward.

File: src/Plutus/Certification/Server.hs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE ImportQualifiedPost #-}
 {-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
```
Comment (shlevy) [inline src/Plutus/Certification/Server.hs:3]: Why was this needed?

File: src/Plutus/Certification/Local.hs [hunk 6]
```diff
@@ -119,6 +126,9 @@ renderSubmitJobField :: RenderFieldJSON SubmitJobField
 renderSubmitJobField (SubmittedRef u) = ("submitted-ref", toJSON $ uriToString id u "")
 renderSubmitJobField (JobID jobId) = ("job-id", toJSON jobId)
 
+renderAbortJobField :: RenderFieldJSON RunningJobField
```
Comment (shlevy) [inline src/Plutus/Certification/Local.hs:129]: Looks like you inlined this?

**Review by shlevy — CHANGES_REQUESTED** [review]

File: src/Plutus/Certification/Server.hs (hunk 5 above)
Comment (bogdan-manole) [reply src/Plutus/Certification/Server.hs:3]: https://github.com/input-output-hk/dapps-certification/blob/286f38211d3c761a24db1878fd890f6ffa149294/src/Plutus/Certification/Server.hs#L61

File: src/Plutus/Certification/Local.hs (hunk 6 above)
Comment (bogdan-manole) [reply src/Plutus/Certification/Local.hs]: done, thanks

File: src/Plutus/Certification/Server.hs (hunk 5 above)
Comment (bogdan-manole) [reply src/Plutus/Certification/Server.hs]: done

File: src/Plutus/Certification/Local.hs (hunk 1 above)
Comment (bogdan-manole) [reply src/Plutus/Certification/Local.hs]: I added the modifications for cleaning the cancellation-map

File: src/Plutus/Certification/Server.hs (hunk 5 above)
Comment (shlevy) [reply src/Plutus/Certification/Server.hs]: Ah no that was fine, just curious.

**Review by shlevy — APPROVED** [review]
